File: extras/Python/cmd_messenger/transport/network/tcp_connection_manager.py
"""``TcpConnectionManager`` — port of C# ``TcpConnectionManager``.

Simple connection manager that retries connecting to a fixed host:port and
optionally runs the watchdog.
"""
from __future__ import annotations

from typing import Optional

from ...cmd_messenger import CmdMessenger
from ...connection_manager import ConnectionManager, DeviceStatus
from .tcp_transport import TcpTransport


class TcpConnectionManager(ConnectionManager):
    """Connection manager for TCP transports."""

    def __init__(
        self,
        tcp_transport: TcpTransport,
        cmd_messenger: CmdMessenger,
        watchdog_command_id: int = 0,
        unique_device_id: Optional[str] = None,
    ) -> None:
        super().__init__(cmd_messenger, watchdog_command_id, unique_device_id)
        if tcp_transport is None:
            raise ValueError("Transport is null.")
        self._tcp_transport = tcp_transport
        # TCP doesn't scan; we always have an explicit endpoint.
        self.device_scan_enabled = False

    def _try_connection(self) -> DeviceStatus:
        self.connected = False
        self._log(1, f"Trying TCP endpoint {self._tcp_transport.host}:{self._tcp_transport.port}.")
        if not self._tcp_transport.connect():
            return DeviceStatus.NOT_AVAILABLE
        status = self._arduino_available(self._tcp_transport.timeout + 250)
        self.connected = status == DeviceStatus.AVAILABLE
        if self.connected:
            self._log(1, f"Connected to {self._tcp_transport.host}:{self._tcp_transport.port}.")
        else:
            self._tcp_transport.disconnect()
        return status

    def _do_work_connect(self) -> None:
        try:
            active = self._try_connection() == DeviceStatus.AVAILABLE
        except Exception:
            active = False
        if active:
            self._connection_found_event()

    def _do_work_scan(self) -> None:
        # No real scan for TCP — just defer to connect.
        self._do_work_connect()

File: extras/TypeScript/src/transport/network/tcpConnectionManager.ts
import { CmdMessenger } from '../../cmdMessenger.js';
import { ConnectionManager, DeviceStatus } from '../../connectionManager.js';
import { TcpTransport } from './tcpTransport.js';

export class TcpConnectionManager extends ConnectionManager {
  constructor(
    private readonly tcpTransport: TcpTransport,
    cmdMessenger: CmdMessenger,
    identifyCommandId = 0,
    uniqueDeviceId?: string
  ) {
    super(cmdMessenger, identifyCommandId, uniqueDeviceId);
    this.deviceScanEnabled = false;
  }

  protected override async doWorkConnect(): Promise<void> {
    const activeConnection = (await this.tryConnection()) === DeviceStatus.Available;
    if (activeConnection) {
      this.connectionFoundEvent();
    }
  }

  protected override async doWorkScan(): Promise<void> {
    await this.doWorkConnect();
  }

  private async tryConnection(): Promise<DeviceStatus> {
    this.connected = false;
    this.log(1, `Trying TCP endpoint ${this.tcpTransport.host}:${this.tcpTransport.port}.`);

    if (!(await this.tcpTransport.connect())) {
      return DeviceStatus.NotAvailable;
    }

    const status = await this.arduinoAvailable(this.tcpTransport.timeout + 250);
    this.connected = status === DeviceStatus.Available;

    if (this.connected) {
      this.log(1, `Connected to ${this.tcpTransport.host}:${this.tcpTransport.port}.`);
    } else {
      await this.tcpTransport.disconnect();
    }

    return status;
  }
}
